mycpp/mylib.h

```diff
@@ -335,11 +347,34 @@ class Str : public gc_heap::Obj {
   }
 
   Str* upper() {
-    NotImplemented();
+    return transform(transform_op::upper);
   }
 
   Str* lower() {
-    NotImplemented();
+    return transform(transform_op::lower);
+  }
+
+  Str* transform(transform_op op)
+  {
+    char* new_buf = static_cast<char*>(malloc(len_));
+    for (int char_index = 0; char_index <= len_; ++char_index)
+    {
+      switch (op)
+      {
+        case transform_op::upper:
```

Comment:
Hm this feels too abstract.  Why branch on every character?

I would just duplicate the `for` loop -- that code will be shorter and faster.  (I use `i` as the index pretty consistently)

Also I started moving stuff into the .cc files in general, although there is no hard rule.  That will avoid the forward declaration of NewStr.

Reply:
As far as being too abstract I'm somewhat indifferent.  Happy to duplicate the for loop if that's what you'd prefer.

Moving it into the .cc files is a good idea.  I'll do that tomorrow morning.

As far as speed goes, I'm pretty confidant this will run at very, very close to the same speed as independent loops.  Branch predictors on constant values are pretty good.  I might write a quick test program tomorrow to confirm/deny this suspicion.